File: oss/clawpipe/swift-sdk/Sources/ClawPipe/Booster.swift
/// Booster — deterministic transforms that skip LLM calls.
///
/// Six built-in rules resolve common prompts locally:
/// math, date, UUID, uppercase, lowercase, reverse.

import Foundation

public struct Booster: Sendable {

    public init() {}

    /// Attempt to resolve `prompt` locally.
    /// Returns the resolved string, or `nil` if no rule matched.
    public func boost(_ prompt: String) -> String? {
        let trimmed = prompt.trimmingCharacters(in: .whitespacesAndNewlines)
        for rule in rules {
            if let result = rule(trimmed) {
                return result
            }
        }
        return nil
    }

    // MARK: - Rule table

    private var rules: [(String) -> String?] {
        [mathRule, dateRule, uuidRule, uppercaseRule, lowercaseRule, reverseRule]
    }

    // MARK: Rule 1 — simple integer arithmetic

    private func mathRule(_ input: String) -> String? {
        let pattern = #"^(?:calculate|compute|what is|evaluate|solve)\s+([\d\s\+\-\*\/\(\)\.]+)$"#
        guard
            let regex = try? NSRegularExpression(pattern: pattern, options: .caseInsensitive),
            let match = regex.firstMatch(
                in: input,
                range: NSRange(input.startIndex..., in: input)
            ),
            let exprRange = Range(match.range(at: 1), in: input)
        else { return nil }

        let expr = String(input[exprRange]).trimmingCharacters(in: .whitespaces)
        // NSExpression supports basic arithmetic
        let nsExpr = NSExpression(format: expr)
        guard let value = nsExpr.expressionValue(with: nil, context: nil) as? NSNumber else {
            return nil
        }
        let d = value.doubleValue
        if d == d.rounded() && !d.isInfinite {
            return String(Int(d))
        }
        return String(d)
    }

    // MARK: Rule 2 — current date

    private func dateRule(_ input: String) -> String? {
        let patterns = [
            #"what(?:'s| is) (?:the )?(?:current )?(?:date|time|day)"#,
            #"(?:today|now|current date)"#,
        ]
        guard input.count < 60 else { return nil }
        for pattern in patterns {
            if input.range(
                of: pattern,
                options: [.regularExpression, .caseInsensitive]
            ) != nil {
                let formatter = ISO8601DateFormatter()
                formatter.formatOptions = [.withInternetDateTime]
                return formatter.string(from: Date())
            }
        }
        return nil
    }

    // MARK: Rule 3 — UUID generation

    private func uuidRule(_ input: String) -> String? {
        let pattern = #"generate\s+(?:a\s+)?uuid"#
        guard input.range(
            of: pattern,
            options: [.regularExpression, .caseInsensitive]
        ) != nil else { return nil }
        return UUID().uuidString.lowercased()
    }

    // MARK: Rule 4 — convert X to uppercase

    private func uppercaseRule(_ input: String) -> String? {
        let pattern = #"convert\s+"(.+)"\s+to\s+uppercase"#
        guard
            let regex = try? NSRegularExpression(pattern: pattern, options: .caseInsensitive),
            let match = regex.firstMatch(in: input, range: NSRange(input.startIndex..., in: input)),
            let wordRange = Range(match.range(at: 1), in: input)
        else { return nil }
        return String(input[wordRange]).uppercased()
    }

    // MARK: Rule 5 — convert X to lowercase

    private func lowercaseRule(_ input: String) -> String? {
        let pattern = #"convert\s+"(.+)"\s+to\s+lowercase"#
        guard
            let regex = try? NSRegularExpression(pattern: pattern, options: .caseInsensitive),
            let match = regex.firstMatch(in: input, range: NSRange(input.startIndex..., in: input)),
            let wordRange = Range(match.range(at: 1), in: input)
        else { return nil }
        return String(input[wordRange]).lowercased()
    }

    // MARK: Rule 6 — reverse X

    private func reverseRule(_ input: String) -> String? {
        let pattern = #"reverse\s+"(.+)""#
        guard
            let regex = try? NSRegularExpression(pattern: pattern, options: .caseInsensitive),
            let match = regex.firstMatch(in: input, range: NSRange(input.startIndex..., in: input)),
            let wordRange = Range(match.range(at: 1), in: input)
        else { return nil }
        return String(String(input[wordRange]).reversed())
    }
}

File: oss/clawpipe/elixir-sdk/lib/clawpipe/booster.ex
defmodule ClawPipe.Booster do
  @moduledoc """
  Deterministic transforms that skip LLM calls entirely.

  Rules applied in order:
  1. Math  — arithmetic expressions
  2. Date  — current date/time queries
  3. UUID  — UUID generation
  4. Uppercase — convert text to uppercase
  5. Lowercase — convert text to lowercase
  6. Reverse — reverse a string
  """

  @math_prefix ~r/^(?:calculate|compute|what is|evaluate|solve)\s+(.+)/i
  @safe_expr   ~r/^[\d\s+\-*\/().,%]+$/
  @date_pat    ~r/what(?:'s| is) (?:the )?(?:current )?(?:date|time|day)|(?:today|now|current date)/i
  @uuid_pat    ~r/generate\s+(?:a\s+)?uuid/i
  @upper_pat   ~r/convert\s+"?(.+?)"?\s+to\s+uppercase/i
  @lower_pat   ~r/convert\s+"?(.+?)"?\s+to\s+lowercase/i
  @reverse_pat ~r/reverse\s+"?(.+?)"?$/i

  @doc """
  Attempt to resolve the prompt without an LLM call.

  Returns `{:ok, text}` when a rule matches, or `:skip` when no rule applies.
  """
  @spec boost(String.t()) :: {:ok, String.t()} | :skip
  def boost(prompt) when is_binary(prompt) do
    trimmed = String.trim(prompt)

    rules = [
      &math_rule/1,
      &date_rule/1,
      &uuid_rule/1,
      &uppercase_rule/1,
      &lowercase_rule/1,
      &reverse_rule/1
    ]

    Enum.reduce_while(rules, :skip, fn rule, _acc ->
      case rule.(trimmed) do
        {:ok, _} = result -> {:halt, result}
        :skip -> {:cont, :skip}
      end
    end)
  end

  # --- Rule 1: Math ---

  defp math_rule(input) do
    case Regex.run(@math_prefix, input, capture: :all_but_first) do
      [expr] ->
        clean = String.trim(expr)

        if Regex.match?(@safe_expr, clean) do
          evaluate_math(clean)
        else
          :skip
        end

      _ ->
        :skip
    end
  end

  defp evaluate_math(expr) do
    # Replace ^ with ** for exponentiation, then evaluate via Code
    sanitized =
      expr
      |> String.replace(",", "")
      |> String.replace("%", "/100")

    try do
      {result, _} = Code.eval_string(sanitized)

      text =
        if is_float(result) and trunc(result) == result do
          Integer.to_string(trunc(result))
        else
          to_string(result)
        end

      {:ok, text}
    rescue
      _ -> :skip
    end
  end

  # --- Rule 2: Date ---

  defp date_rule(input) do
    if Regex.match?(@date_pat, input) and String.length(input) < 60 do
      today = Date.utc_today() |> Date.to_iso8601()
      {:ok, today}
    else
      :skip
    end
  end

  # --- Rule 3: UUID ---

  defp uuid_rule(input) do
    if Regex.match?(@uuid_pat, input) do
      uuid = generate_uuid()
      {:ok, uuid}
    else
      :skip
    end
  end

  defp generate_uuid do
    <<a::32, b::16, _::4, c::12, _::2, d::30, e::48>> = :crypto.strong_rand_bytes(16)

    :io_lib.format(
      "~8.16.0b-~4.16.0b-4~3.16.0b-~2.16.0b~4.16.0b-~12.16.0b",
      [a, b, c, 0b10 * 0x40 + (d >>> 28), d &&& 0x0FFFFFFF, e]
    )
    |> IO.iodata_to_binary()
  end

  # --- Rule 4: Uppercase ---

  defp uppercase_rule(input) do
    case Regex.run(@upper_pat, input, capture: :all_but_first) do
      [text] -> {:ok, String.upcase(text)}
      _ -> :skip
    end
  end

  # --- Rule 5: Lowercase ---

  defp lowercase_rule(input) do
    case Regex.run(@lower_pat, input, capture: :all_but_first) do
      [text] -> {:ok, String.downcase(text)}
      _ -> :skip
    end
  end

  # --- Rule 6: Reverse ---

  defp reverse_rule(input) do
    case Regex.run(@reverse_pat, input, capture: :all_but_first) do
      [text] -> {:ok, String.reverse(text)}
      _ -> :skip
    end
  end
end
